homeassistant/components/shelly/__init__.py

```diff
@@ -27,6 +27,14 @@
 async def async_setup(hass: HomeAssistant, config: dict):
     """Set up the Shelly component."""
     hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][GENERAL] = {}
+    hass.data[DOMAIN][CONFIG_ENTRY] = {}
+    hass.data[DOMAIN][GENERAL][
+        COAP_CONTEXT
+    ] = await aiocoap.Context.create_client_context()
+    hass.data[DOMAIN][GENERAL][UNDO_SHUTDOWN_LISTENER] = hass.bus.async_listen(
+        EVENT_HOMEASSISTANT_STOP, shutdown_listener(hass)
```

Comment:
Don't pass a coroutine as a listener. Rebase on latest `dev` and this will actually raise an exception.

Reply:
Yes I read your earlier comment but I'm not sure what should I change. `shutdown_listener` should be a sync function and call coap context shutdown via `asyncio.run_coroutine_threadsafe`?

Reply:
No. There is a difference between a coroutine and a coroutine function.

```python
# coroutine function
async def bla():
   pass

# generates a coroutine
bla()
```

Reply:
Ok, now I understand, thanks.

Reply:
If I pass the function as a listener
```python
hass.data[DOMAIN][DATA_GENERAL][UNDO_SHUTDOWN_LISTENER] = hass.bus.async_listen(
    EVENT_HOMEASSISTANT_STOP, shutdown_listener
)
```
than `hass.data` is an empty dict in `shutdown_listener`.
Any hint what I'm doing wrong?

Reply:
I moved `shutdown_listener` to `async_setup_entry` and now it has access to `coap_context`.